ui/src/app/views/cohort-edit/component.ts

```diff
@@ -1,68 +1,70 @@
-import {Component, OnDestroy, OnInit} from '@angular/core';
+import {Component} from '@angular/core';
+import {FormBuilder, FormControl, FormGroup, Validators} from '@angular/forms';
 import {ActivatedRoute, Router} from '@angular/router';
-import {Subscription} from 'rxjs/Subscription';
 
 import {ErrorHandlingService} from '../../services/error-handling.service';
 
-import {Cohort, CohortsService} from 'generated';
+import {Cohort, CohortsService, Workspace} from 'generated';
 
-@Component({
-  styleUrls: ['./component.css'],
-  templateUrl: './component.html',
-})
-export class CohortEditComponent implements OnInit, OnDestroy {
-  cohort: Cohort = {
-    name: '',
-    description: '',
-    criteria: '',
-    type: '',
-  };
-  workspaceNamespace: string;
-  workspaceId: string;
-  private sub: Subscription;
-  private loading = false;
+@Component({templateUrl: './component.html'})
+export class CohortEditComponent {
+  /* Properties */
+  loading = false;
+  form: FormGroup;
 
-  constructor(
-      private router: Router,
-      private route: ActivatedRoute,
-      private cohortsService: CohortsService,
-      private errorHandlingService: ErrorHandlingService,
-  ) {}
-
-  ngOnInit(): void {
-    this.sub = this.route.params
-      .do(({ns, wsid}) => {
-        this.workspaceNamespace = ns;
-        this.workspaceId = wsid;
-      })
-      .switchMap(({ns, wsid, cid}) => this.cohortsService.getCohort(ns, wsid, +cid))
-      .subscribe(cohort => this.cohort = cohort);
-  }
+  /* Convenience property aliases */
+  get name(): FormControl         { return this.form.get('name') as FormControl; }
+  get description(): FormControl  { return this.form.get('description') as FormControl; }
+  get cohort(): Cohort            { return this.route.snapshot.data.cohort; }
+  get workspace(): Workspace      { return this.route.snapshot.data.workspace; }
 
-  ngOnDestroy() {
-    this.sub.unsubscribe();
+  constructor(
+    private route: ActivatedRoute,
+    private router: Router,
+    private fb: FormBuilder,
+    private cohortService: CohortsService,
+    private errorHandlingService: ErrorHandlingService,
+  ) {
+    const {cohort} = route.snapshot.data;
```

Comment:
Needed? Can you just use `this.cohort` below?

Reply:
Actually _not_ needed, nice catch.  That is magical - I had though that since `this.cohort` relied on `this.route` it wouldn't be available until _after_ the constructor but I guess Angular does its black magic with the constructor args before the constructor is even called.